pyteal/ast/router.py

```diff
@@ -400,45 +399,85 @@ def _wrap_handler(
                     Approve(),
                 )
 
-    def __add_bare_call(self, oc_actions: OnCompleteActions) -> None:
-        action_type = Expr | SubroutineFnWrapper | ABIReturnSubroutine
-        if oc_actions.is_empty():
-            raise TealInternalError("the OnCompleteActions is empty.")
-        cs_calls, approval_calls = oc_actions._clear_state_n_other_oc()
-        if cs_calls.on_call:
-            on_call = cast(action_type, cs_calls.on_call)
-            wrapped = Router._wrap_handler(False, on_call)
-            self.categorized_clear_state_ast.bare_calls.append(
-                CondNode(Int(1), wrapped)
-            )
-        if cs_calls.on_create:
-            on_create = cast(action_type, cs_calls.on_create)
-            wrapped = Router._wrap_handler(False, on_create)
-            self.categorized_clear_state_ast.bare_calls_create.append(
-                CondNode(Txn.application_id() == Int(0), wrapped)
-            )
-        for oc, approval_bac in approval_calls.items():
-            if approval_bac.on_call:
-                on_call = cast(action_type, approval_bac.on_call)
-                wrapped = Router._wrap_handler(False, on_call)
-                self.categorized_approval_ast.bare_calls.append(
-                    CondNode(Txn.on_completion() == oc, wrapped)
+    def add_method_to_ast(
+        self, method_signature: str, cond: Expr | int, handler: ABIReturnSubroutine
+    ) -> None:
+        walk_in_cond = Txn.application_args[0] == MethodSignature(method_signature)
+        match cond:
+            case Expr():
+                self.conditions_n_branches.append(
+                    CondNode(
+                        walk_in_cond, Cond([cond, self.wrap_handler(True, handler)])
```

Comment:
nit: it'd be slightly more space-efficient to use `Assert` instead of `Cond` because `Cond` uses branching:
```suggestion
                        walk_in_cond, Seq(Assert(cond), self.wrap_handler(True, handler))
```

Reply:
good point, I will make the changes in codegen